[Bug]: Crash when vault_pki_secret_backend_role's "key_usage" is an empty list

### Terraform Core Version

v1.5.7

### Terraform Vault Provider Version

v3.20.1

### Vault Server Version

1.14.2

### Affected Resource(s)

vault_pki_secret_backend_role

### Expected Behavior

Specifiying  `key_usage="[]"` in a `vault_pki_secret_backend_role` resource should either be allowed, or cause an error to be output.

### Actual Behavior

The plugin crashes after a `terraform apply`.

### Relevant Error/Panic Output Snippet

```shell
Stack trace from the terraform-provider-vault_v3.20.1_x5 plugin:

panic: interface conversion: interface {} is nil, not string

goroutine 74 [running]:
github.com/hashicorp/terraform-provider-vault/vault.pkiSecretBackendRol
eCreate({0x24446e8, 0xc000cc5ec0}, 0x1e8b3a0?, {0x1f3dc00, 0xc000c26340
})
        github.com/hashicorp/terraform-provider-vault/vault/resource_pk
i_secret_backend_role.go:454 +0xc65
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).
create(0xc00090ee00, {0x2444720, 0xc000737620}, 0xd?, {0x1f3dc00, 0xc00
0c26340})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.16.0/helper/sch
ema/resource.go:707 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).
Apply(0xc00090ee00, {0x2444720, 0xc000737620}, 0xc000620820, 0xc000886e
00, {0x1f3dc00, 0xc000c26340})
```


### Terraform Configuration Files

N/A

### Steps to Reproduce

1. Create a `vault_pki_secret_backend_role` with key `key_usage` and value `[""]`
2. `terraform apply`
3. Provider crashes

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

Able to reproduce on Linux and Mac (both x86_64)

### References

_No response_

### Would you like to implement a fix?

None

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@ FEATURES:
 
 BUGS:
 * Fix duplicate timestamp and incorrect level messages: ([#2031](https://github.com/hashicorp/terraform-provider-vault/pull/2031))
+* Fix panic when setting `key_usage` to an array of empty string and enable it to unset the key usage constraints: ([#2036](https://github.com/hashicorp/terraform-provider-vault/pull/2036))
 
 ## 3.20.1 (Sep 13, 2023)
 IMPROVEMENTS:

--- a/vault/resource_pki_secret_backend_role.go
+++ b/vault/resource_pki_secret_backend_role.go
@@ -557,7 +557,6 @@ func pkiSecretBackendRoleRead(_ context.Context, d *schema.ResourceData, meta in
 
 	listFields := append(pkiSecretListFields, consts.FieldKeyUsage)
 	// handle TypeList
-	// for _, k := range pkiSecretListFields {
 	for _, k := range listFields {
 		list := expandStringSlice(secret.Data[k].([]interface{}))
 
@@ -565,11 +564,6 @@ func pkiSecretBackendRoleRead(_ context.Context, d *schema.ResourceData, meta in
 			d.Set(k, list)
 		}
 	}
-	// special handling for key_usage because an empty array or array with
-	// empty string means we do not want to specify key usage constraints
-	// if keyUsage, ok := secret.Data[consts.FieldKeyUsage]; ok {
-	// 	d.Set(consts.FieldKeyUsage, keyUsage)
-	// }
 
 	// handle TypeBool
 	for _, k := range pkiSecretBooleanFields {

--- a/vault/resource_pki_secret_backend_role.go
+++ b/vault/resource_pki_secret_backend_role.go
@@ -451,7 +451,14 @@ func pkiSecretBackendRoleCreate(ctx context.Context, d *schema.ResourceData, met
 			ifcList := v.([]interface{})
 			list := make([]string, 0, len(ifcList))
 			for _, ifc := range ifcList {
-				list = append(list, ifc.(string))
+				if ifc != nil {
+					list = append(list, ifc.(string))
+				} else if ifc == nil && k == consts.FieldKeyUsage {
+					// special handling for key_usage because an array with an
+					// empty string means we do not want to specify key usage
+					// constraints
+					list = append(list, "")
+				}
 			}
 
 			if len(list) > 0 {

--- a/vault/resource_pki_secret_backend_role.go
+++ b/vault/resource_pki_secret_backend_role.go
@@ -49,7 +49,6 @@ var pkiSecretListFields = []string{
 	consts.FieldAllowedDomains,
 	consts.FieldAllowedSerialNumbers,
 	consts.FieldExtKeyUsage,
-	consts.FieldKeyUsage,
 }
 
 var pkiSecretBooleanFields = []string{
@@ -448,22 +447,23 @@ func pkiSecretBackendRoleCreate(ctx context.Context, d *schema.ResourceData, met
 	// handle TypeList
 	for _, k := range pkiSecretListFields {
 		if v, ok := d.GetOk(k); ok {
-			ifcList := v.([]interface{})
-			list := make([]string, 0, len(ifcList))
-			for _, ifc := range ifcList {
-				if ifc != nil {
-					list = append(list, ifc.(string))
-				} else if ifc == nil && k == consts.FieldKeyUsage {
-					// special handling for key_usage because an array with an
-					// empty string means we do not want to specify key usage
-					// constraints
-					list = append(list, "")
-				}
-			}
+			list := expandStringSlice(v.([]interface{}))
 
 			if len(list) > 0 {
 				data[k] = list
 			}
+		}
+	}
+	// special handling for key_usage because an empty array or array with
+	// empty string means we do not want to specify key usage constraints
+	if v, ok := d.GetOk(consts.FieldKeyUsage); ok {
+		data[consts.FieldKeyUsage] = expandStringSlice(v.([]interface{}))
+	} else {
+		// check if we are an empty array or null (not set in config)
+		val, _ := d.GetRawConfig().AsValueMap()[consts.FieldKeyUsage]
+		if !val.IsNull() {
+			// value was set as empty array in config
+			data[consts.FieldKeyUsage] = make([]string, 0)
 		}
 	}
 
@@ -555,18 +555,21 @@ func pkiSecretBackendRoleRead(_ context.Context, d *schema.ResourceData, meta in
 	d.Set(consts.FieldBackend, backend)
 	d.Set(consts.FieldName, name)
 
+	listFields := append(pkiSecretListFields, consts.FieldKeyUsage)
 	// handle TypeList
-	for _, k := range pkiSecretListFields {
-		ifcList := secret.Data[k].([]interface{})
-		list := make([]string, 0, len(ifcList))
-		for _, ifc := range ifcList {
-			list = append(list, ifc.(string))
-		}
+	// for _, k := range pkiSecretListFields {
+	for _, k := range listFields {
+		list := expandStringSlice(secret.Data[k].([]interface{}))
 
 		if len(list) > 0 {
 			d.Set(k, list)
 		}
 	}
+	// special handling for key_usage because an empty array or array with
+	// empty string means we do not want to specify key usage constraints
+	// if keyUsage, ok := secret.Data[consts.FieldKeyUsage]; ok {
+	// 	d.Set(consts.FieldKeyUsage, keyUsage)
+	// }
 
 	// handle TypeBool
 	for _, k := range pkiSecretBooleanFields {
@@ -637,15 +640,23 @@ func pkiSecretBackendRoleUpdate(ctx context.Context, d *schema.ResourceData, met
 	data := map[string]interface{}{}
 	for _, k := range pkiSecretListFields {
 		if v, ok := d.GetOk(k); ok {
-			ifcList := v.([]interface{})
-			list := make([]string, 0, len(ifcList))
-			for _, ifc := range ifcList {
-				list = append(list, ifc.(string))
-			}
+			list := expandStringSlice(v.([]interface{}))
 
 			if len(list) > 0 {
 				data[k] = list
 			}
+		}
+	}
+	// special handling for key_usage because an empty array or array with
+	// empty string means we do not want to specify key usage constraints
+	if v, ok := d.GetOk(consts.FieldKeyUsage); ok {
+		data[consts.FieldKeyUsage] = expandStringSlice(v.([]interface{}))
+	} else {
+		// check if we are an empty array or null (not set in config)
+		val, _ := d.GetRawConfig().AsValueMap()[consts.FieldKeyUsage]
+		if !val.IsNull() {
+			// value was set as empty array in config
+			data[consts.FieldKeyUsage] = make([]string, 0)
 		}
 	}
 

--- a/vault/resource_pki_secret_backend_role_test.go
+++ b/vault/resource_pki_secret_backend_role_test.go
@@ -321,6 +321,13 @@ func TestPkiSecretBackendRole_basic(t *testing.T) {
 				ImportState:       true,
 				ImportStateVerify: true,
 			},
+			{
+				Config: testPkiSecretBackendRoleConfig_basic(name, backend, 3600, 7200, `key_usage = [""]`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key_usage.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "key_usage.0", ""),
+				),
+			},
 		},
 	})
 }

--- a/vault/resource_pki_secret_backend_role_test.go
+++ b/vault/resource_pki_secret_backend_role_test.go
@@ -328,6 +328,12 @@ func TestPkiSecretBackendRole_basic(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "key_usage.0", ""),
 				),
 			},
+			{
+				Config: testPkiSecretBackendRoleConfig_basic(name, backend, 3600, 7200, `key_usage = []`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key_usage.#", "0"),
+				),
+			},
 		},
 	})
 }

--- a/website/docs/r/pki_secret_backend_role.html.md
+++ b/website/docs/r/pki_secret_backend_role.html.md
@@ -96,7 +96,7 @@ The following arguments are supported:
 * `key_bits` - (Optional) The number of bits of generated keys
 
 * `key_usage` - (Optional) Specify the allowed key usage constraint on issued
-  certificates. Deafults to `["DigitalSignature", "KeyAgreement", "KeyEncipherment"])`.
+  certificates. Defaults to `["DigitalSignature", "KeyAgreement", "KeyEncipherment"])`.
   Set to an array of empty string `[""]` to unset the default key usage
   constraints.
 

--- a/website/docs/r/pki_secret_backend_role.html.md
+++ b/website/docs/r/pki_secret_backend_role.html.md
@@ -95,7 +95,10 @@ The following arguments are supported:
 
 * `key_bits` - (Optional) The number of bits of generated keys
 
-* `key_usage` - (Optional) Specify the allowed key usage constraint on issued certificates
+* `key_usage` - (Optional) Specify the allowed key usage constraint on issued
+  certificates. Deafults to `["DigitalSignature", "KeyAgreement", "KeyEncipherment"])`.
+  Set to an array of empty string `[""]` to unset the default key usage
+  constraints.
 
 * `ext_key_usage` - (Optional) Specify the allowed extended key usage constraint on issued certificates
 

--- a/website/docs/r/pki_secret_backend_role.html.md
+++ b/website/docs/r/pki_secret_backend_role.html.md
@@ -97,8 +97,7 @@ The following arguments are supported:
 
 * `key_usage` - (Optional) Specify the allowed key usage constraint on issued
   certificates. Defaults to `["DigitalSignature", "KeyAgreement", "KeyEncipherment"])`.
-  Set to an array of empty string `[""]` to unset the default key usage
-  constraints.
+  To specify no default key usage constraints, set this to an empty list `[]`.
 
 * `ext_key_usage` - (Optional) Specify the allowed extended key usage constraint on issued certificates
 

